### ISSUE TITLE
Unet and attention parameters

### ISSUE THREAD

**Issue by YUHANG-Ma** [lucidrains/DALLE2-pytorch#182]
Hi, I am trying to adjust the model parameters to those in DALLE2 paper. 
![image](https://user-images.githubusercontent.com/72591225/176842271-6d08fedf-b6bd-4c15-bb66-823ba72854e9.png)

When I reading the code, I didn't find the attaention resolution and I think "heads channels" in the paper refers to "attn_dim_head" in our code? Could I ask where to set the attaention resolution?

**Comment by lucidrains** [comment]
@YUHANG-Ma hey Yuhang, hope your training has been going well

so the attention layers wasn't accurate to the paper, as I do a full self attention at the lowest resolution (8), and then give all the layers a type of weak linear attention. however, in the latest version v0.15.3, you can now specify, using a tuple, which layer you wish it to have full self attention https://github.com/lucidrains/DALLE2-pytorch/blob/main/dalle2_pytorch/dalle2_pytorch.py#L1349

```python
base_unet = Unet(
  self_attn = (False, True, True, True)
)
```

**Comment by lucidrains** [comment]
@YUHANG-Ma however, if you find that too expensive, you can always turn off the full self attention in lieu of the sparse linear attention (just set `sparse_attn = True`)

**Comment by lucidrains** [comment]
haha, it isn't actually sparse, sort of misnamed it

should rename it to linear attention, but you get the idea. it'll be weaker than the full self attention

**Comment by lucidrains** [comment]
@YUHANG-Ma what kind of results are you seeing so far?

**Comment by YUHANG-Ma** [comment]
Hey lucidrains, thanks for your reply. I use the parameters as follows and with sparse attention. 
dim_mults=(1, 2, 3, 4)
attn_dim_head = 32
attn_heads = 16
resnet_groups = 8
num_resnet_blocks = 2
init_cross_embed_kernel_sizes = (3, 7, 15)
I will try the new version to set the attention resolution and keep updated here.
Also, I want to try to set the parameters the same as the paper. But I didn't understand which one in the paper is related to "attn_dim_head" in the code. So does "attn_heads". I read the paper and it says it uses network the same as GLIDE. In the GLIDE, its head channels is set to 64 but I can't find where I could set this in our code. I thought it could be the "attn_dim_head". But what is "attn_heads" in our code for? 

The thing is, I trained the decoder with about 300M parameters(dim=256). But the result seems not good. The details of the human face is not clear in the sampling examples. I guess there are two possible reasons. The first one is that my dataset(15M) is not that good the other reason I think is that the parameters I set (300M) is far less than those in the paper which are 3.5B. So I want to set the decoder parameters the same as the paper and retrain my model.

**Comment by lucidrains** [comment]
@YUHANG-Ma yea that sounds normal

`attn_heads` refers to the number of attention heads, since all attention is multi-headed these days

**Comment by lucidrains** [comment]
@YUHANG-Ma did you figure out what was wrong on your end for https://github.com/lucidrains/DALLE2-pytorch/issues/155 ?

**Comment by YUHANG-Ma** [comment]
> 
![image](https://user-images.githubusercontent.com/72591225/177008720-c4ff004c-a603-4976-bd79-08097fab79ea.png)

then how about attn_dim_head? which one refers to Head channels in the paper?

**Comment by YUHANG-Ma** [comment]
> @YUHANG-Ma did you figure out what was wrong on your end for #155 ?

yes, the input images and embeddings are not related to each other. I finally fixed it.

**Comment by lucidrains** [comment]
> > 
> 
> ![image](https://user-images.githubusercontent.com/72591225/177008720-c4ff004c-a603-4976-bd79-08097fab79ea.png)
> 
> then how about attn_dim_head? which one refers to Head channels in the paper?

"heads channel" would refer to the `attn_dim_head`

they have it at 64, which is a traditional setting from NLP, but i think 32 is fine for images

**Comment by lucidrains** [comment]
`attn_dim_head` stands for attention dimension per head (or head dimension / channels)

**Comment by YUHANG-Ma** [comment]
OKay, thanks. And how did the paper set the "attn_heads" which is defined in the code? I didn't find it in the paper.

**Comment by lucidrains** [comment]
@YUHANG-Ma they didn't specify, but typically it will be 4, 8, and at most 16

**Comment by lucidrains** [comment]
8 is a good default value

**Comment by lucidrains** [comment]
@YUHANG-Ma you can always just check to see what they set for GLIDE, likely they just kept it the same

**Comment by YUHANG-Ma** [comment]
Okay, I will try to check it in GLIDE. Thanks for your time.

**Comment by YUHANG-Ma** [comment]
Hi lucidrains,
I tried to set the config like this
![image](https://user-images.githubusercontent.com/72591225/177069497-5669d503-0abb-47e2-ade9-6f496957e8f5.png)
But it ends with an error.
![image](https://user-images.githubusercontent.com/72591225/177069526-da2684d0-8d10-4fbb-bb13-11964942f11c.png)
How should I fix it?
